### PR TITLE
Allow FractionalMaxPool 2D and 3D layers to accept 0 dim batch size tensors.

### PR DESCRIPTION
**Pull request by v0dro**
This issue fixes a part of https://github.com/pytorch/pytorch/issues/12013, which is summarized concretely in  https://github.com/pytorch/pytorch/issues/38115.

Allow `FractionalMaxPool` 2D and 3D layers to accept 0 dim batch sizes. Also make some minor corrections to error messages to make them more informative.

### CODE REVIEW

File: aten/src/ATen/native/FractionalMaxPool2d.cpp [hunk 1]
```diff
@@ -30,9 +30,15 @@ TORCH_META_FUNC(fractional_max_pool2d) (
   int64_t poolSizeW = pool_size[1];
 
   int64_t ndims = input.ndimension();
-  TORCH_CHECK(input.numel() > 0 && (ndims == 3 || ndims == 4),
-    "non-empty 3D or 4D (batch mode) tensor expected for input, but got: ",
-    ndims);
+  TORCH_CHECK(ndims == 3 || ndims == 4,
+              "fractional_max_pool2d(): Expected non-empty 3D or 4D tensor with optional 0-dim batch size, but got: ",
```
Comment (jbschlosser) [inline aten/src/ATen/native/FractionalMaxPool2d.cpp:34]: "non-empty" isn't fully accurate now- why not just "Expected 3D or 4D input tensor, but got:" ?

File: aten/src/ATen/native/FractionalMaxPool2d.cpp [hunk 2]
```diff
@@ -30,9 +30,15 @@ TORCH_META_FUNC(fractional_max_pool2d) (
...
+              input.sizes());
+  for (int64_t i = 1; i < ndims; ++i) {
+    TORCH_CHECK(input.size(i) > 0,
+                "fractional_max_pool2d(): Expected input to have non-zero dimensions for non-batch dimensions, but got",
```
Comment (jbschlosser) [inline aten/src/ATen/native/FractionalMaxPool2d.cpp:38]: This might be overly pedantic, but I think the terms "dimension" and "size" should be precisely and consistently used to avoid confusion. Each "dimension" has a "size".

"Expected input to have non-zero size for non-batch dimensions, but got:"

Similar phrasing for error messages throughout

**Review by jbschlosser — COMMENTED** [review]
Thanks for the update! Updates & testing look good- couple minor comments on error message phrasing

File: test/test_nn.py [hunk 3]
```diff
@@ -13300,6 +13300,25 @@ def test_MarginLoss_empty(self, device, dtype):
                 y = torch.ones(10, 0, device=device).type(torch.long)
                 mod(x, y)
 
+    @onlyOnCPUAndCUDA
+    def test_FractionalMaxPool2d_zero_batch(self, device):
+        mod = nn.FractionalMaxPool2d(3, output_ratio=(0.5, 0.5)).to(device)
```
Comment (jbschlosser) [inline test/test_nn.py:13305]: nit: No need to call `to(device)` on the module since it has no parameters or buffers.

File: test/test_nn.py [hunk 4]
```diff
@@ -13300,6 +13300,25 @@ def test_MarginLoss_empty(self, device, dtype):
...
+        inp = torch.ones(0, 16, 50, 32).to(device)
+        self._test_module_empty_input(mod, inp, check_size=False)
+
+        with self.assertRaisesRegex(RuntimeError, "Expected input"):
+            inp = torch.randn(1, 0, 50, 32).to(device)
```
Comment (jbschlosser) [inline test/test_nn.py:13310]: nit: can pass device directly during instantiation e.g. `inp = torch.randn(1, 0, 50, 32, device=device)`

File: aten/src/ATen/native/FractionalMaxPool2d.cpp [hunk 5]
```diff
@@ -30,9 +30,15 @@ TORCH_META_FUNC(fractional_max_pool2d) (
   int64_t poolSizeW = pool_size[1];
 
   int64_t ndims = input.ndimension();
-  TORCH_CHECK(input.numel() > 0 && (ndims == 3 || ndims == 4),
-    "non-empty 3D or 4D (batch mode) tensor expected for input, but got: ",
-    ndims);
+  TORCH_CHECK(ndims == 3 || ndims == 4,
+              "fractional_max_pool2d(): Expected 3D or 4D tensor with optional 0-dim batch size, but got: ",
```
Comment (jbschlosser) [inline aten/src/ATen/native/FractionalMaxPool2d.cpp:34]: nit: I think since this check is only concerned with the number of dims, "fractional_max_pool2d(): Expected 3D or 4D tensor, but got: " is better

File: aten/src/ATen/native/FractionalMaxPool3d.cpp [hunk 6]
```diff
@@ -168,9 +168,15 @@ void fractional_max_pool3d_out_cpu_template(
   int64_t widthDim = 3;
 
   int64_t ndims = input_.ndimension();
-  TORCH_CHECK(input_.numel() != 0 && (ndims == 4 || ndims == 5),
-    "fractional_max_pool3d_out(): non-empty 4D or 5D (batch mode) tensor ",
-    " expected for input, but got: ", ndims);
+  TORCH_CHECK(ndims == 4 || ndims == 5,
+              "fractional_max_pool3d_out(): Expected 4D or 5D tensor with optional 0-dim batch size, but got: ",
```
Comment (jbschlosser) [inline aten/src/ATen/native/FractionalMaxPool3d.cpp:172]: nit: I think since this check is only concerned with the number of dims, "fractional_max_pool2d(): Expected 3D or 4D tensor, but got: " is better

File: aten/src/ATen/native/cuda/FractionalMaxPool3d.cu [hunk 7]
```diff
@@ -167,10 +167,16 @@ void fractional_max_pool3d_out_cuda_template(
 
     int64_t ndims = input.ndimension();
     TORCH_CHECK(
-      input.numel() != 0 && (ndims == 4 || ndims == 5),
-      "fractional_max_pool3d_out_cuda_template(): ",
-      "non-empty 4D or 5D (batch mode) tensor expected for input, but got: ",
-      ndims);
+                ndims == 4 || ndims == 5,
+                "fractional_max_pool3d_out_cuda_template(): ",
+                "Expected 4D or 5D tensor with optional 0-dim batch size, but got: ",
```
Comment (jbschlosser) [inline aten/src/ATen/native/cuda/FractionalMaxPool3d.cu:172]: nit: I think since this check is only concerned with the number of dims, "fractional_max_pool2d(): Expected 3D or 4D tensor, but got: " is better

**Review by jbschlosser — COMMENTED** [review]
Some tiny minor comments then good to go! :)

**Review by jbschlosser — APPROVED** [review]
LGTM! Thanks for the update :)
